File: src/mcp/tools/projects.ts
// Project read tools: projects_list, project_get.

import { z } from 'zod'
import { apiCall } from '../api-client'
import { jsonResponse, type ToolModule, type ToolResponse } from '../shared'
import { toErrorResponse } from '../errors'
import { SlugRef } from '../schemas/common'

const ProjectsListInputSchema = z.object({}).strict()

const ProjectMemberSchema = z.object({
  userId: z.string(),
  role: z.string(),
  user: z.object({ id: z.string(), name: z.string(), email: z.string() }).optional(),
})

const ProjectSchema = z.object({
  id: z.string(),
  slug: z.string(),
  name: z.string(),
  description: z.string().nullable().optional(),
  tags: z.array(z.string()),
  myRole: z.string().optional(),
  environments: z.array(z.object({ name: z.string() })).optional(),
  members: z.array(ProjectMemberSchema).optional(),
}).passthrough()

const ProjectsListOutputSchema = z.object({
  projects: z.array(ProjectSchema),
  count: z.number().int(),
})

const ProjectGetInputSchema = z.object({ slug: SlugRef }).strict()

const ProjectGetOutputSchema = z.object({ project: ProjectSchema })

const LIST_DESCRIPTION = `List envman projects accessible to the authenticated user. Returns only projects where the user is a member (or all projects for SUPER_ADMIN).

ARGS: (none)

RETURNS:
  - projects[]: array of { id, slug, name, description, tags[], myRole, environments[], members[] }
  - count: total number returned

EXAMPLES:
  - "What projects do I have?" → projects_list() → list with slug + name
  - Then use slug as input to other tools like vars_list

ERRORS:
  - 401: Token invalid
  - Network: server unreachable

NOTES:
  - Cached server-side 60s — fresh data may lag briefly after create/update.
  - For details of one project (members, environments with counts) use project_get.`

const GET_DESCRIPTION = `Get full detail of one project: members with their roles, environments with var counts, and the caller's effective role.

ARGS:
  - slug: project slug (string, e.g., "my-app")

RETURNS:
  - project: full Project shape including members[] and environments[] with vars count

EXAMPLES:
  - project_get({ slug: "open-marina" }) → { project: { ... members: [...], environments: [{ name: "dev", _count: { vars: 12 }}, ...]}}

ERRORS:
  - 404: Slug not found — use projects_list to see available slugs
  - 403: No access to this project

NOTES:
  - Use this before vars_list / aliases_list to confirm which environments and members exist.`

interface ProjectsListResponse { projects: unknown[] }
interface ProjectGetResponse { project: unknown }

export const projectsModule: ToolModule = {
  register(server, ctx) {
    server.registerTool(
      'projects_list',
      {
        title: 'List accessible projects',
        description: LIST_DESCRIPTION,
        inputSchema: ProjectsListInputSchema.shape,
        outputSchema: ProjectsListOutputSchema.shape,
        annotations: {
          readOnlyHint: true,
          destructiveHint: false,
          idempotentHint: true,
          openWorldHint: true,
        },
      },
      async (): Promise<ToolResponse> => {
        try {
          const res = await apiCall<ProjectsListResponse>(ctx.cfg, '/api/envman/projects', {
            resource: 'projects',
          })
          return jsonResponse({ projects: res.projects, count: res.projects.length })
        } catch (e) {
          return toErrorResponse(e)
        }
      },
    )

    server.registerTool(
      'project_get',
      {
        title: 'Get project detail',
        description: GET_DESCRIPTION,
        inputSchema: ProjectGetInputSchema.shape,
        outputSchema: ProjectGetOutputSchema.shape,
        annotations: {
          readOnlyHint: true,
          destructiveHint: false,
          idempotentHint: true,
          openWorldHint: true,
        },
      },
      async (args): Promise<ToolResponse> => {
        try {
          const parsed = ProjectGetInputSchema.parse(args)
          const res = await apiCall<ProjectGetResponse>(
            ctx.cfg,
            `/api/envman/projects/${encodeURIComponent(parsed.slug)}`,
            { resource: `Project "${parsed.slug}"`, notFoundHint: 'Use projects_list to see available slugs.' },
          )
          return jsonResponse({ project: res.project as Record<string, unknown> })
        } catch (e) {
          return toErrorResponse(e)
        }
      },
    )
  },
}

File: cli-go/internal/projects/projects.go
// Package projects implements `envman projects`: listing projects and their
// environments via the existing server endpoints.
package projects

import (
	"net/url"

	"github.com/bipprojectbali/envman/cli/internal/api"
	"github.com/bipprojectbali/envman/cli/internal/auth"
)

// Project is one entry of GET /api/envman/projects (subset of fields).
type Project struct {
	Slug        string `json:"slug"`
	Name        string `json:"name"`
	CreatedByID string `json:"createdById"`
	CreatedBy   struct {
		ID    string `json:"id"`
		Name  string `json:"name"`
		Email string `json:"email"`
	} `json:"createdBy"`
	MyRole string `json:"myRole"`
	Count  struct {
		Environments int `json:"environments"`
	} `json:"_count"`
	Environments []struct {
		Name string `json:"name"`
	} `json:"environments"`
}

// List returns all projects visible to the caller.
func List(cfg *auth.Config) ([]Project, error) {
	var resp struct {
		Projects []Project `json:"projects"`
	}
	if err := api.FetchJSON(cfg, "/api/envman/projects", &resp); err != nil {
		return nil, err
	}
	return resp.Projects, nil
}

// Detail is GET /api/envman/projects/:slug (subset: name + environments).
type Detail struct {
	Slug         string `json:"slug"`
	Name         string `json:"name"`
	MyRole       string `json:"myRole"`
	Environments []struct {
		Name       string `json:"name"`
		AccessRole string `json:"accessRole"`
		Count      struct {
			Vars int `json:"vars"`
		} `json:"_count"`
	} `json:"environments"`
}

// Get returns one project's detail, including its environments.
func Get(cfg *auth.Config, slug string) (*Detail, error) {
	var resp struct {
		Project Detail `json:"project"`
	}
	if err := api.FetchJSON(cfg, "/api/envman/projects/"+url.PathEscape(slug), &resp); err != nil {
		return nil, err
	}
	return &resp.Project, nil
}

// FilterMine returns only the projects created by userID.
func FilterMine(list []Project, userID string) []Project {
	out := make([]Project, 0, len(list))
	for _, p := range list {
		if p.CreatedByID == userID {
			out = append(out, p)
		}
	}
	return out
}

// CurrentUserID returns the authenticated user's id (for --me filtering).
func CurrentUserID(cfg *auth.Config) (string, error) {
	var resp struct {
		User struct {
			ID string `json:"id"`
		} `json:"user"`
	}
	if err := api.FetchJSON(cfg, "/api/envman/whoami", &resp); err != nil {
		return "", err
	}
	return resp.User.ID, nil
}
